File: web/src/lib/adapters/schlage.ts
import type { Device, CameraEvent } from "@/types/device";
import type { DeviceAdapter } from "./types";

/**
 * Schlage cloud adapter.
 *
 * pyschlage uses Allegion's cloud API (AWS Cognito for auth, REST for device control).
 * We replicate the same calls here. The Cognito User Pool and Client ID are the
 * same ones the Schlage Home app uses.
 *
 * Auth flow: Cognito USER_PASSWORD_AUTH → access token → REST API calls.
 */

const COGNITO_USER_POOL_ID = "us-east-1_eWzCndKpY";
const COGNITO_CLIENT_ID = "t5836cptp2s1il0v96g7gms9n";
const COGNITO_REGION = "us-east-1";
const API_BASE = "https://api.allegion.yonomi.co/v1";

// Per-instance token cache (keyed by username to support multi-user)
const tokenCache = new Map<string, { token: string; expiresAt: number }>();

async function authenticate(username: string, password: string): Promise<string> {
  const cached = tokenCache.get(username);
  if (cached && Date.now() / 1000 < cached.expiresAt - 60) {
    return cached.token;
  }

  const res = await fetch(
    `https://cognito-idp.${COGNITO_REGION}.amazonaws.com/`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/x-amz-json-1.1",
        "X-Amz-Target": "AWSCognitoIdentityProviderService.InitiateAuth",
      },
      body: JSON.stringify({
        AuthFlow: "USER_PASSWORD_AUTH",
        ClientId: COGNITO_CLIENT_ID,
        AuthParameters: {
          USERNAME: username,
          PASSWORD: password,
        },
      }),
    }
  );

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Schlage auth failed: ${err}`);
  }

  const data = await res.json();
  const result = data.AuthenticationResult;
  tokenCache.set(username, {
    token: result.AccessToken,
    expiresAt: Math.floor(Date.now() / 1000) + result.ExpiresIn,
  });
  return result.AccessToken;
}

async function apiGet(path: string, username: string, password: string): Promise<any> {
  const token = await authenticate(username, password);
  const res = await fetch(`${API_BASE}${path}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Schlage API error: ${res.status}`);
  return res.json();
}

async function apiPut(path: string, body: any, username: string, password: string): Promise<any> {
  const token = await authenticate(username, password);
  const res = await fetch(`${API_BASE}${path}`, {
    method: "PUT",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`Schlage API error: ${res.status}`);
  return res.json();
}

function parseLock(raw: any): Device {
  const isLocked = raw.attributes?.lockState === 1;
  const isJammed = raw.attributes?.lockState === 2;

  return {
    id: raw.deviceId ?? raw.id ?? "",
    name: raw.name ?? "Schlage Lock",
    device_type: "LOCK",
    room: "",
    online: raw.connected ?? true,
    raw_traits: raw.attributes ?? {},
    lock_state: isJammed ? "JAMMED" : isLocked ? "LOCKED" : "UNLOCKED",
    door_state: "UNKNOWN",
    open_percent: 0,
  };
}

export class SchlageAdapter implements DeviceAdapter {
  name = "SchlageAdapter";
  private username: string;
  private password: string;

  constructor(username?: string, password?: string) {
    this.username = username || process.env.SCHLAGE_USERNAME || "";
    this.password = password || process.env.SCHLAGE_PASSWORD || "";
  }

  async listDevices(): Promise<Device[]> {
    const data = await apiGet("/devices", this.username, this.password);
    const devices = Array.isArray(data) ? data : data.devices ?? [];
    return devices
      .filter((d: any) => d.type === "lock" || d.modelName?.includes("Lock"))
      .map(parseLock);
  }

  async getDevice(deviceId: string): Promise<Device | null> {
    try {
      const raw = await apiGet(`/devices/${deviceId}`, this.username, this.password);
      return parseLock(raw);
    } catch {
      return null;
    }
  }

  async lock(deviceId: string): Promise<boolean> {
    try {
      await apiPut(`/devices/${deviceId}`, {
        attributes: { lockState: 1 },
      }, this.username, this.password);
      return true;
    } catch {
      return false;
    }
  }

  async unlock(deviceId: string): Promise<boolean> {
    try {
      await apiPut(`/devices/${deviceId}`, {
        attributes: { lockState: 0 },
      }, this.username, this.password);
      return true;
    } catch {
      return false;
    }
  }

  async getAccessLogs(
    deviceId: string
  ): Promise<Record<string, string>[]> {
    try {
      const data = await apiGet(`/devices/${deviceId}/logs`, this.username, this.password);
      return Array.isArray(data) ? data : data.logs ?? [];
    } catch {
      return [];
    }
  }

  async getRecentEvents(): Promise<CameraEvent[]> {
    return [];
  }
}

File: securehome/adapters/schlage.py
"""Adapter for Schlage Encode smart locks via the pyschlage community library.

Requires: pip install pyschlage

Authentication uses your Schlage Home app credentials (email + password).
These are set via environment variables SCHLAGE_USERNAME and SCHLAGE_PASSWORD.
"""

from __future__ import annotations

import logging
import os

from securehome.models import (
    CameraEvent,
    Device,
    DeviceType,
    LockState,
)

from .base import DeviceAdapter

logger = logging.getLogger(__name__)


def _get_schlage_auth():
    """Authenticate with the Schlage cloud API."""
    import pyschlage

    username = os.environ.get("SCHLAGE_USERNAME", "")
    password = os.environ.get("SCHLAGE_PASSWORD", "")
    if not username or not password:
        raise ValueError(
            "SCHLAGE_USERNAME and SCHLAGE_PASSWORD environment variables are required"
        )

    return pyschlage.Schlage(username=username, password=password)


def _map_lock_state(lock) -> LockState:
    """Map a pyschlage lock's state to our LockState enum."""
    if lock.is_locked:
        return LockState.LOCKED
    if lock.is_jammed:
        return LockState.JAMMED
    return LockState.UNLOCKED


class SchlageAdapter(DeviceAdapter):
    """Controls Schlage Encode locks via the pyschlage library."""

    def __init__(self) -> None:
        self._api = _get_schlage_auth()

    def list_devices(self) -> list[Device]:
        locks = self._api.locks()
        return [self._to_device(lock) for lock in locks]

    def get_device(self, device_id: str) -> Device | None:
        for lock in self._api.locks():
            if lock.device_id == device_id:
                return self._to_device(lock)
        return None

    def lock(self, device_id: str) -> bool:
        for schlage_lock in self._api.locks():
            if schlage_lock.device_id == device_id:
                schlage_lock.lock()
                logger.info("Locked %s", schlage_lock.name)
                return True
        logger.warning("Lock %s not found", device_id)
        return False

    def unlock(self, device_id: str) -> bool:
        for schlage_lock in self._api.locks():
            if schlage_lock.device_id == device_id:
                schlage_lock.unlock()
                logger.info("Unlocked %s", schlage_lock.name)
                return True
        logger.warning("Lock %s not found", device_id)
        return False

    def get_recent_events(self, hours: int = 24) -> list[CameraEvent]:
        # Schlage locks don't produce camera events
        return []

    def get_access_logs(self, device_id: str) -> list[dict]:
        """Return the access history for a specific lock.

        Each entry contains timestamp, user, and action (lock/unlock/code used).
        This is Schlage-specific and not part of the base DeviceAdapter interface.
        """
        for schlage_lock in self._api.locks():
            if schlage_lock.device_id == device_id:
                return schlage_lock.logs()
        return []

    @staticmethod
    def _to_device(lock) -> Device:
        return Device(
            id=lock.device_id,
            name=lock.name,
            device_type=DeviceType.LOCK,
            room="",
            online=lock.is_connected if hasattr(lock, "is_connected") else True,
            lock_state=_map_lock_state(lock),
        )
